src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -57,6 +63,89 @@ def _build_x509_name(backend, x509_name):
     return x509.Name(attributes)
 
 
+def _build_general_name(backend, gn):
+    # dNSName
+    if gn.type == 2:
+        data = backend._ffi.buffer(
+            gn.d.dNSName.data, gn.d.dNSName.length
+        )[:].decode("ascii")
+        value = x509.DNSName(idna.decode(data))
+    # dirName
+    elif gn.type == 4:
+        value = x509.DirectoryName(
+            _build_x509_name(backend, gn.d.directoryName)
+        )
+    # uniformResourceIdentifier
+    elif gn.type == 6:
+        data = backend._ffi.buffer(
+            gn.d.uniformResourceIdentifier.data,
+            gn.d.uniformResourceIdentifier.length
+        )[:].decode("ascii")
+        parsed = urllib_parse.urlparse(data)
+        hostname = idna.decode(parsed.hostname)
+        if parsed.port:
+            netloc = hostname + u":" + str(parsed.port)
```

Comment:
This should be `six.text_type`, not `str` (I don't think it can ever actually fail, since port is always an integer? but it'll be much cleaner)